### PR TITLE
Code coverage improvements - Part II

### PR DESCRIPTION
**Pull request by kunagpal**
* Before:
```
Statements   : 91.11% ( 1168/1282 )
Branches     : 84.53% ( 1082/1280 )
Functions    : 92.66% ( 303/327 )
Lines        : 91.68% ( 1146/1250 )
```

* After:
```
Statements   : 94.46% ( 1211/1282 )
Branches     : 91.16% ( 1166/1279 )
Functions    : 94.51% ( 310/328 )
Lines        : 94.56% ( 1182/1250 )
```

### CODE REVIEW

File: lib/collection/variable-scope.js [hunk 1]
```diff
@@ -141,7 +141,7 @@ _.assign(VariableScope.prototype, /** @lends VariableScope.prototype */ {
      */
     toObject: function (excludeDisabled, caseSensitive) {
         // if the scope has no layers, we simply export the contents of primary store
-        if (!this._layers) {
+        if (_.isEmpty(this._layers)) {
```
Comment (kamalaknn) [inline lib/collection/variable-scope.js]: Is `!this._layers` not sufficient for any cases?

`_.isEmpty()` is 100% slower than `!variable`
screen shot 2017-08-10 at 12 49 19 pm

File: test/unit/request-body.test.js [hunk 2]
```diff
@@ -79,6 +79,85 @@ describe('RequestBody', function () {
...
     });
 
+    describe('.update', function () {
+        it('should use the raw request body by if an invalid mode is specified', function () {
+            var reqData = new RequestBody({
+                mode: 'random',
+                raw: 'This is supposed to be a raw body. Do not cook it.'
+            });
+            expect(reqData).to.be.ok();
+            expect(reqData).to.have.property('mode', 'raw');
```
Comment (kamalaknn) [inline test/unit/request-body.test.js]: Can you add an assertion for the value of body as well?

File: lib/collection/variable-scope.js (hunk 1 above)
Comment (kunagpal) [inline lib/collection/variable-scope.js]: `this._layers` is an empty array by default, so `!` wouldn't be useful here.

File: test/unit/property-list.test.js [hunk 3]
```diff
@@ -720,6 +720,162 @@ describe('PropertyList', function () {
             // should have removed the array of values
             expect(list.reference).to.not.have.property('key1');
         });
+
+        it('should correctly remove an element by direct reference', function () {
+            var entity = new FakeType({ keyAttr: 'key', valeu: 'val' }),
```
Comment (kunagpal) [inline test/unit/property-list.test.js]: Good catch 👍

File: test/unit/property-list.test.js [hunk 4]
```diff
@@ -720,6 +720,162 @@ describe('PropertyList', function () {
...
+            );
+        });
+    });
+
+    describe('.insert', function () {
+        it('should bail out for non-object arguments', function () {
+            var pList = new PropertyList();
+
+            pList.insert();
+            expect(pList.members).to.eql([]);
```
Comment (kunagpal) [inline test/unit/property-list.test.js]: That would also pass for an empty object, the test here is strictly for empty arrays.

File: test/unit/property-list.test.js [hunk 5]
```diff
@@ -720,6 +720,162 @@ describe('PropertyList', function () {
...
+        });
+    });
+
+    describe('.each', function () {
+        var FakeType = function (options) {
+            this.key = options.key;
+            this.value = options.value;
+        };
+
+        it('should handle non-function iterators correctly', function () {
```
Comment (kunagpal) [inline test/unit/property-list.test.js]: The difference comes from the implementation of `_.forEach`, which sanitizes the iterator.

Comment (kamalaknn) [inline test/unit/property-list.test.js]: Ah, okay.

File: test/unit/property-list.test.js (hunk 4 above)
Comment (kamalaknn) [inline test/unit/property-list.test.js]: 👍

File: test/unit/request-body.test.js (hunk 2 above)
Comment (kunagpal) [inline test/unit/request-body.test.js]: Added

File: test/unit/property-list.test.js [hunk 6]
```diff
@@ -720,6 +720,162 @@ describe('PropertyList', function () {
...
+    describe('.eachParent', function () {
+        var FakeType = function (options) {
+            this.key = options.key;
+            this.value = options.value;
+        };
+
+        it('should bail out if the iterator is not a function', function () {
+            var pList = new PropertyList(FakeType, {}, []);
+
+            expect(pList.eachParent.bind(pList)).withArgs('random').to.not.throwError();
```
Comment (kunagpal) [inline test/unit/property-list.test.js]: Done

File: lib/collection/variable.js [hunk 7]
```diff
@@ -131,10 +131,8 @@ _.assign(Variable.prototype, /** @lends Variable.prototype */ {
         }
 
         // set the new type if it is valid and cast the stored value
-        if (typeName) {
-            this.type = typeName;
-            !_nocast && (this.value = this.cast(this.value));
-        }
```
Comment (shamasis) [inline lib/collection/variable.js]: Why remove the if block?

File: lib/collection/variable-scope.js [hunk 8]
```diff
@@ -177,8 +177,8 @@ _.assign(VariableScope.prototype, /** @lends VariableScope.prototype */ {
             i,
             ii;
 
-        // if a variable does not exist in local scope, we search all the layers and return the first occurence.
-        if (!variable && this._layers) {
+        // if a variable does not exist in local scope, we search all the layers and return the first occurrence.
+        if (!(variable || _.isEmpty(this._layers))) {
```
Comment (shamasis) [inline lib/collection/variable-scope.js]: We are creating this._layers on demand. Hence we had no chance of an empty array. Did we?

File: lib/collection/variable-scope.js (hunk 1 above)
Comment (shamasis) [inline lib/collection/variable-scope.js]: That was changed later

File: lib/collection/property.js [hunk 9]
```diff
@@ -163,7 +163,7 @@ _.assign(Property.prototype, /** @lends Property.prototype */ {
             throw Error('Unable to resolve variables. Require a List type property for variable auto resolution.');
         }
 
-        return variables ? variables.substitute(this.toJSON(), overrides) : undefined;
+        return variables.substitute(this.toJSON(), overrides);
```
Comment (shamasis) [inline lib/collection/property.js]: 👍🏼

File: lib/collection/variable.js (hunk 7 above)
Comment (kamalaknn) [inline lib/collection/variable.js]: There is an early return for empty `typeName`. By this line, `typeName` should always be present right?.
